internal/controller/kubecertagent/deployer.go

```diff
@@ -0,0 +1,273 @@
+			agentDeploymentInformer,
+			pinnipedcontroller.SimpleFilterWithSingletonQueue(agentPodConfig.isAgentDeployment),
+			controllerlib.InformerOption{},
+		),
+		// Be sure to run once even to make sure the CredentialIssuer is updated if there are no controller manager
+		// pods. We should be able to pass an empty key since we don't use the key in the sync (we sync
+		// the world).
+		controllerlib.WithInitialEvent(controllerlib.Key{}),
+	)
+}
+
+// Sync implements controllerlib.Syncer.
+func (c *deployerController) Sync(ctx controllerlib.Context) error {
+	// List kube-controller-manager Pods in kube-system.
+	controllerManagerPods, err := c.kubeSystemPodInformer.Lister().List(controllerManagerLabels)
+	if err != nil {
+		return fmt.Errorf("informer cannot list controller manager pods: %w", err)
+	}
+
+	// Find the latest healthy kube-controller-manager Pod.
+	newestControllerManager := newestRunningPod(controllerManagerPods)
+
+	// If there are no healthy controller manager pods, we alert the user that we can't find the keypair via
+	// the CredentialIssuer.
+	if newestControllerManager == nil {
+		c.log.Info("could not find a healthy kube-controller pod", "candidates", len(controllerManagerPods))
+		return issuerconfig.UpdateStrategy(
+			ctx.Context,
+			c.credentialIssuerLocationConfig.Name,
+			c.credentialIssuerLabels,
+			c.pinnipedAPIClient,
+			strategyError(c.clock, constable.Error("could not find a healthy kube-controller-manager pod")),
+		)
+	}
+
+	// Build the expected Deployment based on the kube-controller-manager Pod as a template.
+	c.log.Info("building expected Deployment based on kube-controller-manager Pod", "pod", klog.KObj(newestControllerManager))
+	expectedDeployment := newAgentDeployment(c.agentPodConfig, newestControllerManager)
+
+	// Try to get the existing Deployment, if it exists.
+	existingDeployment, err := c.agentDeploymentInformer.Lister().Deployments(expectedDeployment.Namespace).Get(expectedDeployment.Name)
+	notFound := k8serrors.IsNotFound(err)
+	if err != nil && !notFound {
+		return fmt.Errorf("informer cannot get agent deployments: %w", err)
+	}
+
+	// If the Deployment did not exist, create it and be done.
+	if notFound {
+		c.log.Info("creating new Deployment", "deployment", klog.KObj(expectedDeployment))
+		_, err := c.k8sClient.AppsV1().Deployments(expectedDeployment.Namespace).Create(ctx.Context, expectedDeployment, metav1.CreateOptions{})
+		return err
+	}
+
+	// If the existing Deployment already matches our desired spec, we're done.
+	if apiequality.Semantic.DeepDerivative(expectedDeployment, existingDeployment) {
+		c.log.Info("existing Deployment is already up to date", "deployment", klog.KObj(expectedDeployment))
+		return nil
+	}
+
+	// Otherwise update the spec of the Deployment to match our desired state and update it.
+	c.log.Info("updating existing Deployment", "deployment", klog.KObj(expectedDeployment))
+	expectedDeployment.Spec.DeepCopyInto(&existingDeployment.Spec)
```

Comment:
You are mutating an informer cache item.

Reply:
Doh, refactored this from some code that was reading directly from the client and missed this. Luckily our unit tests in CI caught it too 🎉